### DISCUSSION TITLE
What apps have you developed using Extractor Data? Lets highlight them!

### DISCUSSION THREAD

**Discussion by LukeCroteau** [LukeCroteau/RaidExtractor#46] [Show and tell]
As I'm sitting here removing the links to the website code that exist in this repository, I was thinking it would probably be more useful to instead link to people who have created other apps designed to take in the JSON Data the tool generates!

Drop your links in this thread, with a description of what your tool does, or submit a pull request to add to the Readme file, and lets create a network of examples using this project!

**Comment by cmarabate** [comment]
Great idea!

I haven't made this public yet because it is still in the early stages and I have a lot more plans for it before I am ready to start receiving a bunch of feedback. Lol but here is what I have been working on...
[https://raidchamps.com/mychamps/](https://raidchamps.com/mychamps/)
I don't plan on changing that URL, there is just nothing on my site linking to it yet.

It currently is just a way to pull in all your champs into my site and search through them by name using a search box. They are all sorted as close as I can get them to the order they are in-game with icons to identify which champs are in storage and which are ready to be ranked up with everything looking very similar to how it does in-game.

By the way, this is the project I was referring to when I mentioned that I think I could get the champs to be sorted in the same order they are sorted in-game if each champions power is able to be extracted from the game. Without that, the current sort order is as close as I can get.

The first thing I plan to add is toggles that allow you to easily filter your champs for stuff like "Only show champs ready to rank up", "Show Duplicates", "Show Triplicates" as well as the obvious stuff like Faction, Rarity, Affinity, Level, Grade, & Awaken Level.

Then I plan to figure out a way to allow users to easily share their entire roster with whichever filters they want toggled on. I am hoping to figure out a way to pull this off with a shareable URL of some sort that would somehow preserve the state and localStorage data. But I will cross that bridge when I come to it 🤣. A simpler option that I can already do with a this chrome extension is take a screenshot of the entire page. But I don't want to require an extension and when I shared one of these screenshot with my Clan in Discord, I had to explain to them that they needed to click it and then click Open Original and then click the image the pops up in the browser to zoom in on it so you can see it clearly. So that is not ideal lol.

But now I am just rambling lol, I will leave it at that.

**Reply by LukeCroteau** [reply]
I like how this already looks! There's some cool looking things on this site.
I _love_ being able to see skill modifiers, and doing nerdy things with the knowledge they give me.

I'll add this to the list!

**Reply by cmarabate** [reply]
Thanks 😁, is this the first time you've heard about my site? I am awful at promoting it but it seems to be spreading by word of mouth 🥳.

Did you happen to noticed the Filter Menus on the home page? I think they are the most useful thing on the site but I have a feeling not many people are taking advantage of them. Since you mentioned that you like seeing the skill modifiers, it sounds like you might also find the filter menus useful.

**Reply by cmarabate** [reply]
@LukeCroteau I made it public, so feel free to add it whenever...
[https://raidchamps.com/mychamps](https://raidchamps.com/mychamps)

Thanks!

**Reply by kfishkin** [reply]
hi @cmarabate , it crashes for me - console says "TypeError: Cannot read property 'heroes' of null". I'm guessing maybe you're not checking for a missing cookie?

**Reply by cmarabate** [reply]
Hmm that is strange @kfishkin, you get that error after uploading your JSON file? It is pulling the data from localStorage, not a cookie but yeah it sounds like it is looking for a something that does not exist in the JSON file you uploaded. Did you use the latest version of Luke's Extractor to generate the JSON file? If so, could you send it to me so I can troubleshoot the issue? Thanks for pointing this out!

**Reply by kfishkin** [reply]
ahh, yes, must have been an older version - I just uploaded a newer version and it worked fine. Looks nice! Can I ask where you got the images from? I just stumbled across https://github.com/PatPat1567/RaidShadowLegendsData and have updated 'Brago' to use avators and armor pictures from there.

**Reply by cmarabate** [reply]
Oh good, thanks!

I got a big bulk of them from this Google Drive folder I found a link to on Reddit a long time ago...
[https://drive.google.com/drive/folders/1JIuqgxTl0Zy0jeBFUFfPJQkNlgbxVpPU](https://drive.google.com/drive/folders/1JIuqgxTl0Zy0jeBFUFfPJQkNlgbxVpPU)

But it is not up-to-date at all and was missing a lot of champs from the git go, but it got me started lol. I got the rest of the images either from Ayumilove or by taking screengrabs of their images from the patch notes that Plarium releases on on their forum for champs that Ayumilove had not added the portrait image for yet. But they usually add it shortly after, so if I were you I would scrape them all from Ayumilove.

Somebody on my Discord server just mentioned that repo to me yesterday. I wonder how long it has been a thing lol, it would have saved me a lot of work.

That is exciting that you added the avatars and armor pictures! I have to say I am incredibly impressed by Brago, Are the changes not live on https://laughing-engelbart-62bcb5.netlify.app/ yet? Because I don't see the images yet.

I would love to pick your brain some time on how you got those toggle filters working. I really want to add similar filters to My Champs Roster page but it has turned out to be more difficult than I imagined.

I am also curious if you know how to edit Brago to upload the JSON data to localStorage instead of using cookies. I tried playing with the code from your github repo to pull this off but was unsuccessful. I think it would be much more user friendly if we didn't have to re-upload our JSON file every time we refresh the page.

**Reply by kfishkin** [reply]
I just stumbled across the 'PatPat' repo via Google search, it looks like it was last updated a few weeks ago. I just submitted a PR to add some missing armor sets, we'll see if it's still being maintained.

Yes, the changes are live, e.g. https://i.imgur.com/K2Vv6PG.png shows 'Iron Brago' being displayed on 'Brago'. They only show on the 'Detail' page, not on the champion list page.

As to how the toggle filters work - you can see the source  at https://github.com/kfishkin/brago/blob/main/src/ArtifactSellPage.js, I just run a 'for' loop over every artifact and see if it passes whichever filters are selected. The 'filters' all have the same API contract so they can be stored in an array and easily traversed, e.g.:
```
        this.state.checkers.forEach((checker) => {
          if (this.state.checkedByCheckerId[checker.id]) {
            var why = checker.fn(artifact, numberer);
            if (why) { // this artifact passed this checker
```


It _would_ be nice to save the last JSON to localStorage, thanks for the idea. I've never used localStorage, but I'll look into it.

**Reply by kfishkin** [reply]
oh and PS the somebody on discord was me :)

**Reply by cmarabate** [reply]
> Yes, the changes are live, e.g. https://i.imgur.com/K2Vv6PG.png shows 'Iron Brago' being displayed on 'Brago'. They only show on the 'Detail' page, not on the champion list page.
Oh, that page has never worked for me for some reason. Everytime I select a champ from the select input, the entire page turns white and after I refresh the page I have to load my JSON file again and it does the same thing if I try selecting a champ again...
[https://drive.google.com/file/d/1Zq3tVQ7brHt9w9W4bNst1JzfCHAQR3kU/view?usp=sharing](https://drive.google.com/file/d/1Zq3tVQ7brHt9w9W4bNst1JzfCHAQR3kU/view?usp=sharing)

Thanks for the info about your toggle filters, I will give it another shot.

> It would be nice to save the last JSON to localStorage, thanks for the idea. I've never used localStorage, but I'll look into it.
Sure thing, I think it will make a big difference in how easy it is to use your site. It might even fix your Champion Details page issue, but don't quote me on that lol.

**Reply by kfishkin** [reply]
yikes on the white page! Next time that happens, can you do this (assuming you're using Chrome):

- right-click on the white page and choose the 'inspect' options
- in the top list of tabs, choose 'Console'
- see if there are any error messages printed out.

**Reply by cmarabate** [reply]
![image](https://user-images.githubusercontent.com/12236062/114451550-5bfabc00-9ba5-11eb-9997-8f8262360fe0.png)

**Reply by zerfl** [reply]
Hi @cmarabate,

I found a bug on the website. It looks like there's no publicly available repo and thus I can't create an issue for it. I hope you don't mind posting it here instead:

If you visit https://raidchamps.com/mychamps/ with an empty localStorage (incognito will do) the react application will throw an exception at some point because `item` is empty:

./src/components/UseLocalStorageFileUploader.js:121
```
if (! item) {
  setItem( AllChampsDataKey, JSON.stringify(placeholderData) )
}
  const data = JSON.parse(item);
  const champs = data.heroes;
  //const artifact = data.artifacts;



  const sortByRarity = ['Common', 'Uncommon', 'Rare', 'Epic', 'Legendary',]
  const customSort = ({data, sortByRarity, sortField}) => {
  const sortByObject = sortByRarity.reduce((obj, item, index) => {
    return {
      ...obj,
      [item]: index
    }
  }, {})
  return data.sort((a, b) => sortByObject[a[sortField]] - sortByObject[b[sortField]])
}
```
Once you reload the page everything appears to be working fine. From what I've gathered you're doing the following:

1. retrieve the content of localStorage (and store it in `item`) // L25
2. check for empty localStorage // L121
3. fill with placeholder data if empty // L122
4. access hero data using `item`. // L124

Imho, the problem lies between step 3 and 4. While step 3 does successfully write the placeholder data to the storage, accessing it using `item` at this point will not work because it still holds the old content from step 1 instead of the placeholder data.

HTH,

**Reply by cmarabate** [reply]
Thanks @zerfl, I was aware of this issue but I didn't think there was much I could do about it since HellHades Raid Optimizer has the same exact issue and everyone has gotten used to having to refresh their page whenever it only loads the white screen. His actually is worse because I get that white screen whenever I haven't used his optimizer in awhile whereas mine only has that issue the first time that you visit the page and it should not happen again afterward unless you clear out your browsers localStorage.

Thanks for your tips about where the problem lies, I will look into it further. Do you happen to also have a solution to the problem?

**Reply by raidchamps** [reply]
@zerfl I came up with a really simple solution to this problem. I just added a line of code to make the page automatically reload after setting the placeholder data. Thanks for bringing this up!

**Reply by zerfl** [reply]
I am no expert in React but couldn't you just do this instead:

```
if (!item) {
  item = JSON.stringify(placeholderData); // make the placeholder data available for the initial session
  setItem(AllChampsDataKey, item); // store placeholder data in localstorage
}
```

**Reply by cmarabate** [reply]
Thanks but I tried your code and it still requires the page to be reloaded. But I think my solution is all that is needed anyways.

**Comment by kfishkin** [comment]
hi Luke, thanks for the idea. The URL on my tool is gibberish (that's what ya get with free hosting), best to do it like this:
[Brago ](https://laughing-engelbart-62bcb5.netlify.app/)- Basic Raid Gear Observer. Recommends artifacts to sell and enchant.

**Reply by cmarabate** [reply]
Domain names are cheap 😜. Netlify makes it really easy to setup a custom domain name with their free plan, you just have to buy one from any domain registrar or from Netlify themselves.

**Reply by LukeCroteau** [reply]
> hi Luke, thanks for the idea. The URL on my tool is gibberish (that's what ya get with free hosting), best to do it like this:
> [Brago ](https://laughing-engelbart-62bcb5.netlify.app/)- Basic Raid Gear Observer. Recommends artifacts to sell and enchant.

I updated the link in the readme to reflect this. Thanks for updating it!

**Comment by papedre** [comment]
Hello, 
Actually working on a tool that will extract the data from your json. 
It will try (with some parameters) to estimate a score for each artifacts. 
Then it will extract these data to an Excel file.
The artifacts will be ordered like in the game. 
Each line will give you an advice for the artifact (sell, keep, upgrade).
The advice isn't based on single artifact but based on all the artifacts you got. 

The objective is to be able to quickly decide what to sell and to upgrade. 

I still need to tune my scoring method, and create a config file for all the parameters.

**Reply by redafcb10** [reply]
Go for it dude!! I need some help cleaning my vault. If I can help in any way let me know (testing for example)

**Reply by kfishkin** [reply]
@papedre that does sound cool! Can I talk you into putting it in with 'Brago'? That already has a UI that shows all the artifacts and possible reasons to sell, we could add a column with your algorithm's recco.

**Reply by cmarabate** [reply]
Sounds cool, I like that the advice will be based on all of our artifacts. That is huge! Will it also give advice that takes the set into account as well? I know certain sets are only good with specific stats, so that would be a big help as well.

**Reply by papedre** [reply]
Anyone knows exactly how the artifacts are ordered in the game ??

**Reply by LukeCroteau** [reply]
> Anyone knows exactly how the artifacts are ordered in the game ??

They're sorted in the order of:
Star Value
Level
Rarity
Internal ID

**Comment by kfishkin** [comment]
Hi all, hope it's ok to post this here - just wanted to let folks know that thanks to some suggestions by @cmarabate I've made some major improvements to [Brago ](https://main--laughing-engelbart-62bcb5.netlify.app/)- you can now use it to examine champions, not just artifacts. So e.g. you can say things like "show me champions with an unfilled accessory", "show me champions with dupes",or my personal favorite, "show me champions with gear that's >= 2 stars lower than they are:"
![champions_inferior](https://user-images.githubusercontent.com/6618823/114798843-2b548700-9d4b-11eb-91e2-5173ae8e6170.png)

**Reply by cmarabate** [reply]
It has come a long way in a very short amount of time. Great work, I am very impressed! I am going to start using this every day, it is so useful. I already told you this but just so everyone who watches these posts knows, there is now a link to @kfishkin's site in the "Links" menu on [RaidChamps.com](https://raidchamps.com) under "Tools"...
![image](https://user-images.githubusercontent.com/12236062/115126743-7a92f580-9f9f-11eb-8299-b53d6460a841.png)

**Comment by kfishkin** [comment]
hi @zerfl and @LukeCroteau ,
Just wanted to say thanks for adding the Arena and Great Hall info into the JSON file.
I think it's in exactly the right format, BTW.
I've already modified 'Brago' to read the Arena info part, should have the Great Hall in by tomorrow.

thanks again!

**Comment by cmarabate** [comment]
I created another one using the newly extracted Shard Mercy Data!
![image](https://user-images.githubusercontent.com/12236062/115932237-8a588100-a45a-11eb-9433-ede79e384b49.png)
After pulling my shards...
![image](https://user-images.githubusercontent.com/12236062/115932263-95abac80-a45a-11eb-8dba-f4c5ef65f150.png)

Try it out at [https://raidchamps.com/mymercy](https://raidchamps.com/mymercy)

I am also now pulling in all the tags into each of our champs on [https://raidchamps.com/mychamps](https://raidchamps.com/mychamps)...
![image](https://user-images.githubusercontent.com/12236062/115933262-83cb0900-a45c-11eb-8fc5-6505612a32aa.png)


Thanks for all the upgrades to the Extracter, this are really starting to get exciting!

**Reply by WickedShadows** [reply]
Not an app but really close! I have developed a mega spreadsheet with every champ in it and all thier stats and so on and off that a built a team builder spreadsheet that uses your output in parts. Check it out, let me know what you think!

https://wickedwallz.blogspot.com/2021/06/raid-shadow-legends-team-builder-plus.html

**Comment by kfishkin** [comment]
hi everybody, just FYI I've enhanced '[Brago](https://laughing-engelbart-62bcb5.netlify.app/)' so that it can show you your total stats, and you can sort by them as well, e.g. below I'm sorting my 'magic' champs by HP:
![health_desc](https://user-images.githubusercontent.com/6618823/116140257-ec4bfd00-a68b-11eb-908f-f28dec34211a.png)
